src/utilities/keyForFragmentSpread.ts

```diff
@@ -0,0 +1,23 @@
+import type { FragmentSpreadNode } from '../language/ast.js';
+import { print } from '../language/printer.js';
+
+/**
+ * Create a key that uniquely identifies common fragment spreads.
+ * Treats the fragment spread as the source of truth for the key: it
+ * does not bother to look up the argument definitions to de-duplicate default-variable args.
+ *
+ * Using the fragment definition to more accurately de-duplicate common spreads
+ * is a potential performance win, but in practice it seems unlikely to be common.
+ */
+export function keyForFragmentSpread(fragmentSpread: FragmentSpreadNode) {
+  const fragmentName = fragmentSpread.name.value;
+  const fragmentArguments = fragmentSpread.arguments;
+  if (fragmentArguments == null || fragmentArguments.length === 0) {
+    return fragmentName;
+  }
+
+  const printedArguments: Array<string> = fragmentArguments
+    .map(print)
+    .sort((a, b) => a.localeCompare(b));
+  return fragmentName + '(' + printedArguments.join(',') + ')';
```

Comment:
> Should this just be strictly disallowed in the first place similar to how repeated fields need to have equivalent arguments?

Yes, IMO: we should be treating fragment spreads with the same name *exactly like* fields with the same name that need to be able to merge. If we don't it greatly increases the complexity of field merging validation, and likely of the executor.

If you're only changing a variable that isn't actually used within one spread's context, technically the executor ought to be able to handle it. But I think the devex is worse to allow that case than to just make it invalid.

> If so, then we'd need to know the result of fragmentArgumentSubstitutions as an input to determining this key. ...AorB discussion...

This is a good test case to add for OverlappingFieldsCanBeMergedRule, and it should fail (but likely doesn't as-is, great catch!). I think we might need to add a local-variable-scope indicator to the key for validation (though at execution time we should always have actualized values). Good call that the keying function can't *quite* be shared as-is.

Reply:
Oh! I really like this. It's constraining, but in a way that's far more simplifying than restricting. I think if you apply this then you don't need "fragment key" as a concept at all!

I imagine the validation rule either akin to  or part of `OverlappingFieldsCanBeMergedRule` which ensures any two fragment spreads of the same name in the same scope would always produce exactly the same set of fragment variable values (aka: always has the same source of values). I think most of the logic you added is approaching that, but I had just misunderstood that.

In that case, you can assume that for a valid query you only need to spread the first instance of any fragment with a given name since the above rule guarantees that any fragment you encounter later with the same name would provide identical results.

Because you can assume that, you should be able to just keep the existing logic of using the fragment name instead of computing a key during `collectFields`